### ISSUE TITLE
[RFC]: Automate Speculative Decoding

### ISSUE THREAD

**Issue by LiuXiaoxuanPKU** [vllm-project/vllm#4565]
### Motivation.

Speculative Decoding is a crucial feature for reducing latency, currently supported by vLLM (credit to @cadedaniel !). However, when deploying Speculative Decoding in real online LLM serving systems that use continuous batching, improvements are not always observed. Paradoxically, under conditions of high request rates or low speculation accuracy, latency may actually increase.

We propose to address these issues. We want to intelligently determines the optimal speculation length for each request, ranging from zero (no speculation) to multiple tokens. This determination is based on the concept of goodput, which reflects the current observed load across the entire system, thus allowing for most effective speculative execution.

The method is designed for versatility, compatible with various speculative decoding styles, from traditional, model-based approaches to model-free methods such as prompt lookup and tree-style decoding. This innovation builds on recent research by the vLLM team. We plan to release the detailed paper shortly.

### Proposed Change.

**Milestone 1:** Implement a mechanism to disable speculative decoding (proposed length = verified length = 0), allowing users to manually decide when to cease speculative decoding. Based on prior empirical studies, we can initiate this process by monitoring the running_queue size. Speculative decoding will be suspended for incoming requests once the running_queue exceeds a predefined threshold. Cody will assist with this implementation, thanks @comaniac!

**Milestone 2:** Dynamically determine the proposed length for speculative decoding. We will utilize runtime information, such as batch size, in conjunction with profiled parameters like token acceptance rate and the comparative costs of running the draft versus the target model. This approach allows us to adjust the proposed length in real-time, optimizing performance based on current system conditions.

**Milestone 3:** Eliminate reliance on pre-profiled parameters and gather necessary information directly from runtime. We will collect data such as the token acceptance rate and the execution times for both the draft and target models from previous steps. This data will then be integrated into the goodput calculation, allowing for a more dynamic and responsive system configuration

### Any Other Things.

1. We will implement modifications after the scheduler allocates the slots, which may result in some memory inefficiency. For instance, if num_lookahead_slots is set to 5 but the proposed length is only 3, then 2 slots would go unused.
2. Currently, we support proposing lengths at the batch level, meaning all requests within the same batch share the same proposed length. In the future, we could consider supporting more finely grained proposed lengths as needed.

**Comment by cade** [comment]
As much as I would _love_ to take credit for bringing Speculative Decoding to vLLM, I'm relatively certain the praise belongs to @cadedaniel. 😁

**Comment by YuCheng-Qi** [comment]
@LiuXiaoxuanPKU 
Thanks a lot for the super helpful info! I am very interested in the Dynamic Speculative Decoding mentioned above, and also found that the existing vllm framework cannot load two LLM models at the same time (one of which is used as a draft model and the other as a target model). I now have 3 questions for you:
1. Can the vllm framework now support loading two LLM models at the same time?
2. Will Dynamic Speculative Decoding-related functions be developed on the vllm framework?
3. If the above two functions are supported by the vllm community, when will they be implemented?

**Comment by LiuXiaoxuanPKU** [comment]
> @LiuXiaoxuanPKU Thanks a lot for the super helpful info! I am very interested in the Dynamic Speculative Decoding mentioned above, and also found that the existing vllm framework cannot load two LLM models at the same time (one of which is used as a draft model and the other as a target model). I now have 3 questions for you:
> 
> 1. Can the vllm framework now support loading two LLM models at the same time?
> 2. Will Dynamic Speculative Decoding-related functions be developed on the vllm framework?
> 3. If the above two functions are supported by the vllm community, when will they be implemented?

Thanks for the interest! 
1. If the two models are used for speculative decoding, yes vllm can already support that. Take a look at this [worker](https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/spec_decode_worker.py), which contains a draft worker and a target worker. The draft worker is responsible for loading and executing the draft model, while the target work is used for the target model. 
2. Yes, DSD will be integrated into vllm.
3. Currently, we are in the process of optimizing speculative decoding performance because DSD will not be interesting of the native speculative decoding performance is not good. Once we think the native speculative decoding performance is reasonable, we will add DSD on top of it quickly. 
4. For the timeline, since DSD is very light weighted, milestone2 (pre collect some system numbers, support limited models such as llama-7b, llama-70b) can be done within one week. milestone3 is to fully automate the speculation, which will take longer, 1-2 month.

**Comment by YuCheng-Qi** [comment]
@LiuXiaoxuanPKU Thanks for your response, and best wishes to you as well!

**Comment by wooyeonlee0** [comment]
@LiuXiaoxuanPKU Thank you (and other co-workers) for the great work!
I checked the SmartSpec paper on arxiv and slides in the recent meetup, and it looks great :)
I'm looking forward to seeing it in the vllm repo.
Please let me know, if there's anything I can help with. 👍

**Comment by brotherchen** [comment]
amazing works ! I would like to know whether lookahead decoding technology is in the planning of vllm, because it can speed up reasoning without the need for an additional draft model or any additional training.

**Comment by cadedaniel** [comment]
@brotherchen more info [here](https://github.com/vllm-project/vllm/issues/1742) on lookahead decoding in vLLM. currently not being worked on as far as I know.

**Comment by smart-lty** [comment]
> > @LiuXiaoxuanPKU Thanks a lot for the super helpful info! I am very interested in the Dynamic Speculative Decoding mentioned above, and also found that the existing vllm framework cannot load two LLM models at the same time (one of which is used as a draft model and the other as a target model). I now have 3 questions for you:
> > 
> > 1. Can the vllm framework now support loading two LLM models at the same time?
> > 2. Will Dynamic Speculative Decoding-related functions be developed on the vllm framework?
> > 3. If the above two functions are supported by the vllm community, when will they be implemented?
> 
> Thanks for the interest!
> 
> 1. If the two models are used for speculative decoding, yes vllm can already support that. Take a look at this [worker](https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/spec_decode_worker.py), which contains a draft worker and a target worker. The draft worker is responsible for loading and executing the draft model, while the target work is used for the target model.
> 2. Yes, it will be integrated into vllm.
> 3. Currently, we are in the process of optimizing speculative decoding performance because dynamically adjusting it will not be interesting if the native speculative decoding performance is not good. Once we think the native speculative decoding performance is reasonable, we will add our method on top of it quickly.  I am not sure how long this step will take, @cade might have more context here.
> 4. For the timeline, since our method is very light weighted, milestone2 (pre collect some system numbers, support limited models such as llama-7b, llama-70b) can be done within one week. milestone3 is to fully automate the speculation, which will take longer, 1-2 month.

@LiuXiaoxuanPKU  Very interesting work! Our [recent work](https://arxiv.org/pdf/2408.11850) on speculative decoding reveals that executing the draft model and the target model in parallel can achieve an adaptive draft length, which can significantly improve the speculative decoding performance.  I would like to know whether the draft worker and the target worker can execute in parallel within vllm?

**Comment by LiuXiaoxuanPKU** [comment]
> > > @LiuXiaoxuanPKU Thanks a lot for the super helpful info! I am very interested in the Dynamic Speculative Decoding mentioned above, and also found that the existing vllm framework cannot load two LLM models at the same time (one of which is used as a draft model and the other as a target model). I now have 3 questions for you:
> > > 
> > > 1. Can the vllm framework now support loading two LLM models at the same time?
> > > 2. Will Dynamic Speculative Decoding-related functions be developed on the vllm framework?
> > > 3. If the above two functions are supported by the vllm community, when will they be implemented?
> > 
> > 
> > Thanks for the interest!
> > 
> > 1. If the two models are used for speculative decoding, yes vllm can already support that. Take a look at this [worker](https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/spec_decode_worker.py), which contains a draft worker and a target worker. The draft worker is responsible for loading and executing the draft model, while the target work is used for the target model.
> > 2. Yes, it will be integrated into vllm.
> > 3. Currently, we are in the process of optimizing speculative decoding performance because dynamically adjusting it will not be interesting if the native speculative decoding performance is not good. Once we think the native speculative decoding performance is reasonable, we will add our method on top of it quickly.  I am not sure how long this step will take, @cade might have more context here.
> > 4. For the timeline, since our method is very light weighted, milestone2 (pre collect some system numbers, support limited models such as llama-7b, llama-70b) can be done within one week. milestone3 is to fully automate the speculation, which will take longer, 1-2 month.
> 
> @LiuXiaoxuanPKU Very interesting work! Our [recent work](https://arxiv.org/pdf/2408.11850) on speculative decoding reveals that executing the draft model and the target model in parallel can achieve an adaptive draft length, which can significantly improve the speculative decoding performance. I would like to know whether the draft worker and the target worker can execute in parallel within vllm?

Currently no. The draft model and target model are executed sequentially. I image the asynchronous execution will be a big change for the current vllm's arch, but any discussion and contribution is welcomed!

**Comment by TechxGenus** [comment]
This hf blog looks great, can it be easily integrated into vllm?
https://huggingface.co/blog/dynamic_speculation_lookahead

**Comment by gopalsarda** [comment]
@LiuXiaoxuanPKU Amazing work! I have 2 questions regarding the implementation for `Milestone 2: Dynamically determine the proposed length for speculative decoding`
1. Are you planning to provide an abstraction so that individual spec decoding algorithms (EAGLE/Medusa) can implement their own policy for dynamically determining the proposed length or have a stopping criteria? This abstraction would help implement features like what is proposed in [EAGLE-2](https://arxiv.org/pdf/2406.16858) for dynamically adjusting the shape of the draft tree based on the context. This would be helpful even for top-1 proposer.
2. I was curious how would determining proposed length using runtime information like token acceptance rate work for cases where the same endpoint has to serve requests for different tasks (with different acceptance rates). Wouldn't acceptance rate from requests from one task adversely affect the proposal length for requests for another task?

**Comment by sfc-gh-zhwang** [comment]
cc @sfc-gh-zhwang

**Comment by kuangdao** [comment]
m

**Comment by fluctlux** [comment]
@LiuXiaoxuanPKU  I read the Turbo-spec/Smart spec and believe it's a very effective optimization for real-world production environments. I'm curious about its subsequent development. Has this feature already been integrated into vLLM, or is it currently under development?

**Comment by ekagra-ranjan** [comment]
Hi folks, I have been working on this on and off for a while and finally I have raised a PR https://github.com/vllm-project/vllm/pull/32374. Pls have a look!
